### PR TITLE
Add pytest_splinter_screenshot_dir hook

### LINKED ISSUE #119: Overriding splinter_screenshot_dir fixture doesn't work with pytest-xdist
When pytest-xdist is used, the screenshot directory location is stored during pytest_configure. As a result, overriding splinter_screenshot_dir in a conftest.py file has no effect.

### PR DESCRIPTION
**Pull request by jsfehler**
Fixes #119 

- The splinter_screenshot_dir fixture now pulls its value from this hook instead of the config option directly.
- For non-xdist runs, this hook can be used instead of the splinter_screenshot_dir fixture. It will make no real difference.
- For xdist runs, this hook allows the screenshot directory to be overridden from a conftest file instead of only as a command-line option

### CODE REVIEW

File: pytest_splinter/hooks.py [hunk 1]
```diff
@@ -0,0 +1,6 @@
+import pytest  # NOQA
+
+
+def pytest_splinter_screenshot_dir(config):
```
Comment (bubenkoff) [inline pytest_splinter/hooks.py]: not sure if hooks are meant for this
even your default implementation tell that something is broken: why it's an empty function?

Comment (bubenkoff) [reply pytest_splinter/hooks.py]: new hook is not documented as well

File: pytest_splinter/plugin.py [hunk 2]
```diff
@@ -225,7 +236,7 @@ def splinter_make_screenshot_on_failure(request):
 @pytest.fixture(scope='session')  # pragma: no cover
 def splinter_screenshot_dir(request):
     """Browser screenshot directory."""
-    return os.path.abspath(request.config.option.splinter_screenshot_dir)
+    return request.config.hook.pytest_splinter_screenshot_dir(config=request.config)[0]
```
Comment (bubenkoff) [inline pytest_splinter/plugin.py]: why does it take the first?, this is strange

File: pytest_splinter/plugin.py [hunk 3]
```diff
@@ -617,14 +628,12 @@ class SplinterXdistPlugin(object):
 
     """Plugin class to defer pytest-xdist hook handler."""
 
-    def __init__(self, screenshot_dir):
-        """Initialize the SplinterXdistPlugin with the required configuration."""
-        self.screenshot_dir = screenshot_dir
-
     def pytest_testnodedown(self, node, error):
         """Copy screenshots back from remote nodes to have them on the master."""
+        screenshot_dir = node.config.hook.pytest_splinter_screenshot_dir(config=node.config)[0]
```
Comment (bubenkoff) [inline pytest_splinter/plugin.py]: same

File: pytest_splinter/hooks.py (hunk 1 above)
Comment (jsfehler) [reply pytest_splinter/hooks.py]: Other hook implementations I've seen use an empty function to define the hook. Is this not correct?

File: pytest_splinter/plugin.py (hunk 2 above)
Comment (jsfehler) [reply pytest_splinter/plugin.py]: I misunderstood how hooks store values. I've added the hookspec decorator with firstresult=True to address this.

File: pytest_splinter/hooks.py [hunk 4]
```diff
@@ -0,0 +1,9 @@
+from pluggy import HookspecMarker
+
+hookspec = HookspecMarker("pytest")
+
+
+@hookspec(firstresult=True)
+def pytest_splinter_screenshot_dir(config):
```
Comment (bubenkoff) [inline pytest_splinter/hooks.py]: still, please document the hook in the readme

**Comment by youtux** [comment]
I am very much against using hooks for configuration. IMO configuration should really be in the pytest.ini or via command line arguments. I think we should also remove the configuration via fixtures.

**Comment by bubenkoff** [comment]
@youtux I agree hooks are not meant for this, but fixtures as a config are not bad, you know it from the paylogic right?

File: pytest_splinter/hooks.py (hunk 4 above)
Comment (jsfehler) [reply pytest_splinter/hooks.py]: Done

**Comment by jsfehler** [comment]
@youtux I don't disagree. Fixtures aren't perfect for configuration either. This situation shows the limitation of fixtures to modify config data.

Should the pytest-splinter plugin allow configuration to be set and modified in python code? If so, there's only the choice of hooks or fixtures.

In this case, the splinter_screenshot_dir fixture is used by the rest of the plugin. This hook acts as an in-between for the command line argument and the fixture. The advantage of keeping the fixture is easily retrieving the value inside a test, or for use by other fixtures.

Off the top of my head, a different pattern would be:
- Replace fixtures with hooks
- Expose the results as a single fixture with a dict of the results, ie:
```
@pytest.fixture(scope='session')
def splinter_configuration(request):
    config = {
        'splinter_screenshot_dir': '...',
        'splinter_headless': True,
        etc, etc
    }
    return config
```

Of course, overriding such a fixture would be pointless.

**Comment by youtux** [comment]
I would propose another approach then, assuming it would work (I cannot check it by my self at least for the next week, so maybe you can help me).
I would move all the configuration to the ini file, and remove the configuration fixtures.
To solve the problem with having an ad-hoc configuration for the slaves, I would instead use the hook from pytest-xdist that is invoked on slave initialization. In that hook, I would modify the pytest config object, upgrading the value of the configuration key that you want to change (splinter_screenshot_dir in this case).
This should fix the issue, and it’s a user space hook implementation, so pytest-splinter doesn’t have to be aware of this edge case. Maybe we can just mention the trick in the documentation.

**Pull request closed without merging**